cerberus/client.py

```diff
@@ -324,6 +324,169 @@ def update_sdb(self, sdb_id, owner=None, description=None, user_group_permission
+        return self._get_file(secure_data_path, version).content
+
+    def get_file_versions(self, secure_data_path, limit=None, offset=None):
+        """
+        Get versions of a particular file
+        This is just a shim to get_secret_versions
+
+        secure_data_path -- full path to the file in the safety deposit box
+        limit -- Default(100), limits how many records to be returned from the api at once.
+        offset -- Default(0), used for pagination.  Will request records from the given offset.
+        """
+
+        return self.get_secret_versions(secure_data_path, limit, offset)
+
+    def _get_all_file_version_ids(self, secure_data_path, limit=None):
+        """
+        Convenience function that returns a generator that will paginate over the file version ids
+        secure_data_path -- full path to the file in the safety deposit box
+        limit -- Default(100), limits how many records to be returned from the api at once.
+        """
+
+        offset = 0
+        # Prime the versions dictionary so that all the logic can happen in the loop
+        versions = {'has_next': True, 'next_offset': 0}
+        while (versions['has_next']):
+            offset = versions['next_offset']
+            versions = self.get_file_versions(secure_data_path, limit, offset)
+            for summary in versions['secure_data_version_summaries']:
+                yield summary
+
+    def _get_all_file_versions(self, secure_data_path, limit=None):
+        """
+        Convenience function that returns a generator yielding the contents of all versions of
+        a file and its version info
+
+        secure_data_path -- full path to the file in the safety deposit box
+        limit -- Default(100), limits how many records to be returned from the api at once.
+        """
+        for secret in self._get_all_file_version_ids(secure_data_path, limit):
+            yield {'secret': self.get_file_data(secure_data_path, version=secret['id']),
+                   'version': secret}
+
+    def list_files(self, secure_data_path, limit=None, offset=None):
+        """Return the list of files in the path.  May need to be paginated"""
+
+        # Make sure that limit and offset are in range.
+        # Set the normal defaults
+        if not limit or limit <= 0:
+            limit = 100
+
+        if not offset or offset < 0:
+            offset = 0
+
+        payload = {'limit': str(limit), 'offset': str(offset)}
+
+        # Because of the addition of versionId and the way URLs are constructed, secure_data_path should
+        #  always end in a '/'.
+        secure_data_path = self._add_slash(secure_data_path)
+        secret_resp = requests.get(self.cerberus_url + '/v1/secure-files/' + secure_data_path,
+                                   params=payload, headers=self.HEADERS)
+        throw_if_bad_response(secret_resp)
+        return secret_resp.json()
+
+    def put_file(self, secure_data_path, filename, filehandle, content_type=None):
```

Comment:
This function requires file names in two places: path and filename. Is there a reason to having this duplication?

Reply:
From reading the API docs, it looks like filename is required as part of the metadata and isn't derived from the path. 

Form a quick test where I omit the filename, Cerberus will throw an error.  (Granted it could also be my code doing something wrong when it makes the request).

I also tested just suppling the path without the filename, and then supplying the filename in the metadata, but that also didn't work. (once again, it could be my code doing the wrong thing).

Basically it looks like the filename in the path is just the path, whereas the filename in the header information is supplied by the uploader. 

I could try and write the code so that is takes a path without a filename, and then uses the filename parameter to complete the path for you.  But I feel that might be confusing since writing secrets into Cerberus follows a similar  kind of pattern.

Reply:
@splintercat That is partially correct. The filename is required in the HTTP form multipart, but is ignored by Cerberus. What matters is the filename that is after the last slash in the path. I realize the API docs may be misleading in this regard, so let me know if you have ideas on how they can be made more clear.

I think the best way to avoid confusion is to ensure that the full path (including the filename, at the end) is provided to the `put_file` method. Then the filename can be parsed from the given path.

Let me know what you think.

Reply:
I can see about adding that in. 
It does make sense that those two parts should match.